Replace deprecated DOMNodeInserted listener with MutationObserver

Thibault-MTZ opened the pull request:
Replace deprecated DOMNodeInserted listener with MutationObserver
- Resolve Chrome incompatibility when adding collection items or other dynamic nodes  
- Restore cross-browser support (Chrome, Mozilla, etc.)  
- Future-proof DOM mutation handling

Thibault-MTZ committed:
Update injector.html.twig

Replace deprecated DOMNodeInserted listener with MutationObserver
- Resolve Chrome incompatibility when adding collection items or other dynamic nodes  
- Restore cross-browser support (Chrome, Mozilla, etc.)  
- Future-proof DOM mutation handling

bobvandevijver commented:
Thanks for your PR! Can you please update your changes so that only what needs to be changed is actually changed?

Thibault-MTZ committed:
Update injector.html.twig

Thibault-MTZ commented:
@bobvandevijver 
I just adjusted as requested, I think it will be better.

[](https://github.com/bolt/article/pull/44/commits/e2ed90a4c4e317aedb8f9ada921ec1cdc8be4f38)

bobvandevijver commented on templates/injector.html.twig:
```diff
@@ -23,12 +23,33 @@ function initArticles() {
+                    if (node.querySelector && node.querySelector('.collection-item')) {
+                        initArticles();
+                        return;
+                    }
```
This feels double, especially when compared with your previous PR.

bobvandevijver commented on templates/injector.html.twig:
```diff
@@ -23,12 +23,33 @@ function initArticles() {
+                    if (
+                        node.nodeType === Node.ELEMENT_NODE &&
+                        node.classList.contains('collection-item')
+                    ) {
```
```suggestion
                    if (node.nodeType === Node.ELEMENT_NODE && node.classList.contains('collection-item')) {
```

bobvandevijver commented on templates/injector.html.twig:
```diff
@@ -23,12 +23,33 @@ function initArticles() {
+    };
+
+    const articleObserver = new MutationObserver((mutations) => {
+        mutations.forEach((mutation) => {
```
Let's use the for..of loop, to keep this equal to redactor.

bobvandevijver commented on templates/injector.html.twig:
```diff
@@ -23,12 +23,33 @@ function initArticles() {
+        childList: true,
+        subtree:   true
+    };
+
```
```suggestion
```

bobvandevijver commented on templates/injector.html.twig:
```diff
@@ -23,12 +23,33 @@ function initArticles() {
+        });
+    });
+
+    articleObserver.observe(document.body, config);
```
```suggestion
    articleObserver.observe(document.body, {
        childList: true,
        subtree:   true
    });
```

bobvandevijver commented on templates/injector.html.twig:
```diff
@@ -23,12 +23,33 @@ function initArticles() {
+    articleObserver.observe(document.body, config);
 
-initArticles();
+    document.addEventListener('DOMContentLoaded', initArticles);
```
I do not think this change is needed.

Thibault-MTZ committed:
Update templates/injector.html.twig

Co-authored-by: Bob van de Vijver <bobvandevijver@users.noreply.github.com>

Thibault-MTZ committed:
Update injector.html.twig

Thibault-MTZ committed:
Update injector.html.twig

Use MutationObserver, rename observer

- Replace `DOMNodeInserted` with `MutationObserver` for better performance.
- Rename observer to `observerArticle` to prevent naming conflicts.
- Initialize editor on page load and for dynamically added nodes.

Thibault-MTZ commented:
@bobvandevijver Is it okay for you regarding the last commit?
Sorry for the inconvenience, I've just started using Git and GitHub again ;)

bobvandevijver committed:
Update injector.html.twig

bobvandevijver committed:
Update injector.html.twig

bobvandevijver committed:
Update injector.html.twig

bobvandevijver commented:
Thank you!
